packages/core/src/pubsub/pubsub.ts

```diff
@@ -0,0 +1,86 @@
+import { Observable, EMPTY, pipe, of, from, Subscription, UnaryFunction } from 'rxjs';
+import { deserialize, PubsubMessage, serialize } from './pubsub-message';
+import { IpfsApi } from '@ceramicnetwork/common';
+import { map, catchError, mergeMap, withLatestFrom } from 'rxjs/operators';
+import { IncomingChannel, filterOuter, IPFSPubsubMessage } from './incoming-channel';
+import { DiagnosticsLogger, ServiceLogger } from '@ceramicnetwork/logger';
+import { Memoize } from 'typescript-memoize';
+
+/**
+ * Deserialize incoming message in an internal observable that does not emit if error happens.
+ * Log a successfully deserialized message.
+ *
+ * @param peerId$ - own IPFS node peer id
+ * @param pubsubLogger - logger that dumps a successfully deserialized message
+ * @param topic - IPFS pubsub topic we listen
+ */
+function ipfsToPubsub(
+  peerId$: Observable<string>,
+  pubsubLogger: ServiceLogger,
+  topic: string,
+): UnaryFunction<Observable<IPFSPubsubMessage>, Observable<PubsubMessage>> {
+  return pipe(
+    withLatestFrom(peerId$),
+    mergeMap(([incoming, peerId]) =>
+      of(incoming).pipe(
+        map((incoming) => {
+          const message = deserialize(incoming);
+          const logMessage = { ...incoming, data: message };
+          delete logMessage.key;
+          delete logMessage.signature;
+          pubsubLogger.log({ peer: peerId, event: 'received', topic: topic, message: logMessage });
+          return message;
+        }),
+        catchError(() => EMPTY),
+      ),
+    ),
+  );
+}
+
+/**
+ * Receive and publish messages to IPFS pubsub.
+ */
+export class Pubsub extends Observable<PubsubMessage> {
+  constructor(
+    private readonly ipfs: IpfsApi,
+    private readonly topic: string,
+    private readonly resubscribeEvery: number,
+    private readonly pubsubLogger: ServiceLogger,
+    private readonly logger: DiagnosticsLogger,
+  ) {
+    super((subscriber) => {
+      const incoming$ = new IncomingChannel(ipfs, topic, resubscribeEvery, pubsubLogger, logger);
+
+      incoming$
+        .pipe(filterOuter(this.peerId$), ipfsToPubsub(this.peerId$, pubsubLogger, topic))
+        .subscribe(subscriber);
+    });
+  }
+
+  @Memoize()
+  private get peerId$() {
```

Comment:
ditto my earlier comment.  Can this just be a const member rather than a memoized accessor function?

Reply:
Do you find https://gist.github.com/ukstv/1d5943744a31dbe7151f763c3bf5ee71 clearer?

Memoize and getter explicitly state "this is lazy evaluation, executed once, it is just a function". 
In gist, I have to recall ordering of JS object instantiation, and it is a bit frightening. But maybe this is only my quirk.

Reply:
Ah okay, then the issue is that `super()` has to be the first line of the constructor, so then we can't set `this.peerId$` until after the code that references it, which looks weird unless you really understand how the subscription is lazy.  Making it a memoized function would make it work even if the callback passed to `super()` was run immediately.

I'd actually probably still vote for the form from the Gist and just add a comment above where `this.peerId$` is initialized saying that this is safe because the above code from `super` is run lazily.  But I do understand better now why you did it this way so if you prefer it this way that's fine by me

Reply:
Okay.